tests/test_web_poet_rules.py

```diff
@@ -0,0 +1,1173 @@
+    assert ProductDuplicateDeepDependencyPage.to_item_call_count == 1
+
+    # calling the actual page objects should still work
+
+    PageObjectCounterMixin.clear()
+    item, deps = yield crawl_item_and_deps(ProductDuplicateDeepDependencyPage)
+    assert item == MainProductD(
+        name="(with duplicate deep item dependency) product name",
+        item_dependency=ItemDependency(name="item dependency"),
+        main_product_b_dependency=MainProductB(
+            name="(with item dependency) product name",
+            item_dependency=ItemDependency(name="item dependency"),
+        ),
+        first_main_product_c_dependency=MainProductC(
+            name="(with deep item dependency) product name",
+            item_dependency=ItemDependency(name="item dependency"),
+            main_product_b_dependency=MainProductB(
+                name="(with item dependency) product name",
+                item_dependency=ItemDependency(name="item dependency"),
+            ),
+        ),
+        second_main_product_c_dependency=MainProductC(
+            name="(with deep item dependency) product name",
+            item_dependency=ItemDependency(name="item dependency"),
+            main_product_b_dependency=MainProductB(
+                name="(with item dependency) product name",
+                item_dependency=ItemDependency(name="item dependency"),
+            ),
+        ),
+    )
+    assert_deps(deps, {"page": ProductDuplicateDeepDependencyPage})
+    PageObjectCounterMixin.assert_instance_count(1, ItemDependencyPage)
+    PageObjectCounterMixin.assert_instance_count(1, ProductWithItemDepPage)
+    PageObjectCounterMixin.assert_instance_count(1, ProductDeepDependencyPage)
+    PageObjectCounterMixin.assert_instance_count(1, ProductDuplicateDeepDependencyPage)
+    assert ItemDependencyPage.to_item_call_count == 1
+    assert ProductWithItemDepPage.to_item_call_count == 1
+    assert ProductDeepDependencyPage.to_item_call_count == 1
+    assert ProductDuplicateDeepDependencyPage.to_item_call_count == 1
+
+
+@attrs.define
+class ChickenPage:
+    name: str
+    other: str
+
+
+@attrs.define
+class EggPage:
+    name: str
+    other: str
+
+
+@handle_urls(URL)
+@attrs.define
+class ChickenDeadlockPage(ItemPage[ChickenPage]):
+    other_injected: EggPage
+
+    @field
+    def name(self) -> str:
+        return "chicken"
+
+    @field
+    def other(self) -> str:
+        return self.other_injected.name
+
+
+@handle_urls(URL)
+@attrs.define
+class EggDeadlockPage(ItemPage[EggPage]):
+    other_injected: ChickenPage
+
+    @field
+    def name(self) -> str:
+        return "egg"
+
+    @field
+    def other(self) -> str:
+        return self.other_injected.name
+
+
+@inlineCallbacks
+def test_page_object_with_item_dependency_deadlock(caplog) -> None:
+    """Items with page objects which depend on each other resulting in a deadlock
+    should have a corresponding error raised.
+    """
```

Comment:
Or maybe even a case when both pages use items as dependencies.

Reply:
Good one! The current test only checks if both POs have deadlock due to both of them requiring items.

I've added a new variation which is one PO requires items while the other requires a PO. https://github.com/scrapinghub/scrapy-poet/pull/88/commits/33a03910fd60b183ceaff79a435ab4b2411cc644

If two POs both need each others, it'd be hard to make this mistake since Python would err out one PO as not being defined. If they're defined in separate modules and imports one another, Python would also raise cyclic imports.

Reply:
Would it make sense to check the for items as well?

```py
@handle_urls(URL)
@attrs.define
class ChickenDeadlockPage(ItemPage[ChickenItem]):
    egg: EggItem

    @field
    def name(self) -> str:
        return "chicken"

    @field
    def other(self) -> str:
        return self.egg.name


@handle_urls(URL)
@attrs.define
class EggDeadlockPage(ItemPage[EggItem]):
    chicken: ChickenItem  # depend on item, not on a page

    @field
    def name(self) -> str:
        return "egg"

    @field
    def other(self) -> str:
        return self.chicken.name
```

Reply:
Yes, we have it in https://github.com/scrapinghub/scrapy-poet/pull/88/files#diff-a5ce1f1bde36b639be5be17fed1ed35ffb018c1fefcf4544c0e3fac120ca58ebR1314-R1339.